File: pset1/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

int main(void)
{
  long card_number;
  do {
    card_number = get_long("Number: ");
  } while (card_number < 0);

  int length = floor(log10(card_number)) + 1;

  long temp = card_number;
  int checksum = 0;
  for (int i = 0; i < length; i++) {
    int digit = temp % 10;
    temp /= 10;

    // every other number starting from second to last
    if (i % 2 == 1) {
      digit *= 2;
      checksum += digit >= 10 ? (digit % 10) + 1 : digit;
    } else {
      // all other numbers
      checksum += digit;
    }
  }

  string type = "";
  // checksum should end with 0
  if (checksum % 10 > 0) {
    type = "INVALID";
  } else {
    // determine type based on first two numbers
    int prefix = card_number / pow(10, length - 2);

    if (length == 15 && (prefix == 34 || prefix == 37)) {
      type = "AMEX";
    } else if (length == 16 && (prefix > 50 && prefix < 56)) {
      type = "MASTERCARD";
    } else {
      prefix /= 10;
      if ((length == 13 || length == 16) && prefix == 4) {
        type = "VISA";
      } else {
        type = "INVALID";
      }
    }
  }

  printf("%s\n", type);

  return 0;
}

File: pset6/credit/credit.py
from cs50 import get_int
import math


def main():
    while True:
        card_number = get_int('Card number: ')
        if card_number > 0:
            break

    print(card_type(card_number))


def card_type(number):
    temp = number
    checksum = 0
    length = int(math.log10(number)) + 1

    for i in range(length):
        digit = temp % 10
        temp //= 10

        if i % 2 == 1:
            digit *= 2
            checksum += digit if digit < 10 else ((digit % 10) + 1)
        else:
            checksum += digit

    if (checksum % 10) > 0:
        type = 'INVALID'
    else:
        pref = number // 10 ** (length - 2)

        if length == 15 and (pref == 34 or pref == 37):
            type = 'AMEX'
        elif length == 16 and (pref > 50 and pref < 56):
            type = 'MASTERCARD'
        else:
            pref //= 10
            if (length == 13 or length == 16) and pref == 4:
                type = 'VISA'
            else:
                type = 'INVALID'

        return type


if __name__ == "__main__":
    main()
